small modifications to support 2 things. 1) fetching domains parameters ...

lalford opened the pull request:
Hello foreman crew, thanks for all the work you all have put into this system. My colleague and I are making extensive use of it. One of our primary use cases is the external node classifier where we can maintain application configuration data for different environments (modeled as foreman domains).

This requires maintaining a lot of domain parameters, and as such I wanted two things that weren't supported.

1) programmatic access to domain parameters for scripted crud/copy operations. This is what the change to the domains controller did for me.

2) display sort order of existing domain parameters on the edit form so it's easier to scroll through the list to find what I'm interested in. This is what the changes to the domain model and involved views did for me.

It may not be a priority for you all, but the changes are mostly cosmetic and I think would be nice to have in the distribution. It is worth noting that I did this work directly on our installed copy of 1.1-stable, so this diff should backport into those versions without issue. It is also worth noting that I haven't tested this change on any newer versions of foreman, so those of you with more knowledge than me should of course audit whether or not this is the best way to solve the problem.

lalford committed:
small modifications to support 2 things. 1) fetching domains parameters along with the domains. 2) ordering the list of domain parameters by name in the view.

ohadlevy commented:
@lalford thanks, I'll fire on the the tests... I think that much of the API for domain parameters has greately improved in apiv2.. @isratrade would you mind reviewing?

ohadlevy commented:
[test]

domcleal commented:
ping @isratrade.  It'd be great to solve the domain parameter sorting generically, as this is an issue I've heard a few times before (particularly on the host edit page).

isratrade commented:
@lalford, I will look into this soon.

ohadlevy commented:
@isratrade ping?

isratrade commented on app/controllers/domains_controller.rb:
```diff
@@ -20,7 +20,7 @@ def new
   def show
     respond_to do |format|
-      format.json { render :json => @domain }
+      format.json { render :json => @domain.to_json({:methods => [:domain_parameters]}) }
```
@lalford, respond_to json in the UI controllers are soon to be depreciated.  I reccomend using API v2 (which is not the default yet) to retrieve domain parameters.  Just pass version=2 in the header like this:

curl -u admin:secret -H 'Accept:application/json,version=2' http://0.0.0.0:3000/api/domains/123/parameters

isratrade commented:
@lalford, to sort the parameters by name, I suggest that all you need to do is all this line to app/models/parameter.rb

``` ruby
  default_scope lambda { order("parameters.name") }
```

This will also affect the parameter sort order on the edit forms for hosts, hostgroups, and operating systems.  If for some unknown reason, we we only want to the domain parameters to sort by name, then copy this same line into app/models/domain_parameter.rb

No need to modify the views for sorting.

lalford commented:
@isratrade, thanks for the comments. I'm not a rails developer and was unaware of the default_scope method, that sounds handy and I'll give it a try when I get some time. Provided that sorts all parameters for all views that's much better than my domain hack. I'll also try the version 2 api to see if fetching domain parameters works without the controller change. One of these days I need to rewrite my utility scripts to use you all's ruby client as well, I was brand new to foreman when I first wrote those and just threw together a quick client of my own.

Thanks again for the continued feedback.

isratrade commented:
@lalford, great. Glad to hear that this fix works for you.  If others in the community think that parameters should be sorted asc by default, then we will merge it. Just update this pull request and let's see.  Welcome to the Foreman community.

lalford committed:
Revert "small modifications to support 2 things. 1) fetching domains parameters along with the domains. 2) ordering the list of domain parameters by name in the view."

This reverts commit 8c49031e3a68acbf85cdac1d8f56c4a954686b1f.

Reverting the solution to parameter sorting in favor of a cleaner implementation using default_scope that will be forthcoming.

lalford committed:
cleaner and more versatile parameter sorting implementation using default_scope in the parameter model

lalford commented:
@isratrade, I tested the default_scope solution you proposed and it does work as advertised. The host inherited parameters list on the host editor still isn't sorted, but those are computed and more difficult to handle. I definitely like the cleaner implementation with default_scope and having that also work for things like hostgroups (which I hadn't originally supported). I reverted the previous commit and added your solution instead, I also went ahead and removed the domain parameters change in the domain_controller. I haven't gotten a chance to try the version 2 api yet, but parameter sorting and that change are separate issues so I'll open a new request in the future if I still need something along those lines. Thanks!

domcleal commented:
Sorry this slipped through for a few weeks @lalford.  Merged as 189f9f3 for Foreman 1.3.0, with a small addition for the inherited parameters bit :)  Thanks for the patch!
